### ISSUE TITLE
This approach works!

### ISSUE THREAD

**Issue by rgbkrk** [nteract/snakestagram#7]
I created an environment in the same way as the Travis setup:

```bash
conda create -p /tmp/snakestagram --copy -y -q python=3 pandas scikit-learn ipykernel
tar czf /tmp/snakestagram.tar.gz ./snakestagram
```

Later, I extracted this tarball and created a `kernel.json` that references it:

```json
{
 "argv": [
  "/Users/kylek/Desktop/snakestagram/bin/python",
  "-m",
  "ipykernel",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3 (Snakestagram)",
 "language": "python"
}
```

Kernel works in both `jupyter console` and `nteract`. Happy day!

Since we can also install R into a conda environment, we could include an R kernel as well. :smile:

**Comment by rgbkrk** [comment]
Note: our environment with pandas + scikit-learn + ipykernel is a whopping 649 MB unpacked.

If the environment is only ipykernel, it still weighs in at 100MB unpacked.

```
kylek@flixbook /tmp$ du -smh minikernel
100M	minikernel
kylek@flixbook /tmp$ du -smh snakestagram/
648M	snakestagram/
```

This is why I would dub our current nteract as "nteract-lite". An nteract that bundles a kernel is a bit more thick. However, many of the dependencies that make this big are the same across the PyData stack (libmkl + numpy + scipy + pandas and sklearn):

```
/tmp/snakestagram/lib$ du -smh * | gsort -h | tail
 17M	libmkl_sequential.dylib
 18M	libmkl_tbb_thread.dylib
 26M	libmkl_core.dylib
 31M	libmkl_intel_thread.dylib
 33M	libmkl_mc.dylib
 35M	libmkl_mc3.dylib
 37M	libmkl_avx.dylib
 49M	libmkl_avx2.dylib
 52M	libmkl_avx512.dylib
202M	python3.6
```

Deeper look at `site-packages` inside of the python directory:

```
/tmp/snakestagram/lib/python3.6/site-packages$ du -smh * | gsort -h | tail
2.5M	pytz
2.9M	zmq
3.0M	tornado
5.3M	IPython
5.4M	pygments
7.2M	pip
 16M	numpy
 20M	sklearn
 44M	pandas
 47M	scipy
```

**Comment by betatim** [comment]
(We get python3.6? Woah!)

We can build several flavours mini, full stack, R.

I don't have much experience with this but remember something about stripping binaries and such stuff? Could that help with making things smaller? The environment we deliver probably comes with all sorts of caveats (will `pip` work??) so we could add some more like "we stripped everything down to the minimum so that it just about still works, so expect weird stuff when you start adding new packages"? Not sure if it is a good idea, but it is an idea.

**Comment by rgbkrk** [comment]
I think it's better to have working and a known quantity rather than lean. If someone wants lean they can install conda themselves and manage their own environment.

**Comment by minrk** [comment]
Switching to conda-forge (#13) makes a difference because it uses openblas instead of mkl:

```
$ du -hs /tmp/snakestagram
414M	/tmp/snakestagram
```

It *should* be safe to strip static libraries, if you don't care about compiling new things in the env:

```
$ find /tmp/snakestagram -name 'lib*.a' -exec rm '{}' \;
$ du -hs /tmp/snakestagram
317M	/tmp/snakestagram
```

What's the ultimate goal of the bundled kernel?

**Comment by rgbkrk** [comment]
The ultimate goal is to be able to ship applications with a full python stack. With nteract, we'd be able to have a big edition that a student or analyst can download that lets them get started quickly.

**Comment by minrk** [comment]
I'm *slightly* surprised that this works, because my understanding was that conda envs are not relocatable - that is, you cannot zip them up and move them somewhere else.

For instance: every script in `bin/` will have the absolute path of the original directory as its executable.

You know about [conda constructor](https://github.com/conda/constructor)? It may not be precisely what you want, but it lets you build a proper installer based on any conda packages+channels that will do the extra work needed to relocate conda packages. It might make more sense to bundle one of these in your nteract.app and add its install command as a first-run step.

*ix installers are a shell script, so can be run with `bash my-conda.sh -b -p internal-directory`. I'm not sure about the Windows installer, but it might also be able to run in the background.

**Comment by rgbkrk** [comment]
That is correct, all the scripts still have the absolute path of the original directory as its executable. This is also discussed in https://github.com/nteract/snakestagram/issues/9#issuecomment-272349315. From @quasiben:

> All other installed executables will have a header like: #!/tmp/snakestagram/bin/python. When you relocated that folder that probably won't be valid anymore. You probably know this but i'll state it for completeness, you can still execute that particular script by issuing the full path to python and the path to that script as `/other/path/snakestagram/bin/python /other/path/snakestagram/bin/2to3`

----

> You know about conda constructor?

Not until now, thank you!

**Comment by betatim** [comment]
conda constructor looks like what we should be using. For it to work its magic the installer would have to be executed when the user first runs `nteract`. Right now there is no "install" step to start using nteract which is nice. How much of a bodge is it to try and detect if this is the first run of nteract and execute the installer created by conda constructor?

The obvious/clean solution would be to create an explicit install step :-/

**Comment by rgbkrk** [comment]
There's an installer on windows and mac (though it's the DMG kind), it's the built in kind as part of electron builder though.

The current approach does work for an individual kernel (which also means we can do upgrades). I'd like to see us upload the binaries that we currently build to GitHub releases to start trying it out.

**Comment by betatim** [comment]
Can you unjumble your last sentence? 🔀

**Comment by minrk** [comment]
> For it to work its magic the installer would have to be executed when the user first runs nteract

Could it work on first request of the Python kernel? Or even lazily *download and install* the Python kernel on first request of the Python kernel?

**Comment by rgbkrk** [comment]
Whoops! That's definitely jumbled. 😨 

I'm saying that we should try out what we have - it works with ipykernel as is.
